cds-plugin.js

```diff
@@ -30,196 +33,220 @@ const addSideEffects = (actions, flag, element) => {
-        const parentElement = parentElements?.[parentEntityName] ?? {};
+        const parentElement = parentElements?.[parentEntityName] ?? {}
         if (parentElement.type === 'cds.Composition') {
-          const isCompositionEntity = parentElement.target === entity.name;
+          const isCompositionEntity = parentElement.target === entity.name
           // add parent information in the current entity
           if (isCompositionEntity) {
             csn.definitions[entity.name][hasParent] = {
               associationName: name,
               entityName: target
-            };
+            }
           }
-          return isCompositionEntity;
+          return isCompositionEntity
         }
-      });
+      })
     }
-  });
+  })
+}
+
+
+
+const _enhanced = 'sap.changelog.enhanced'
+const namespace = 'sap.changelog'
+const Changes = 'ChangeView'
+const changes = {
+  type: 'cds.Association', target: namespace +'.' + Changes,
+  on: [ {ref:['changes','entityKey']}, '=', /* filled in below */ ],
+  cardinality: { max:'*' },
+}
+const UIFacet = {
+  $Type  : 'UI.ReferenceFacet',
+  Target : 'changes/@UI.PresentationVariant',
+  Label  : '{i18n>ChangeHistory}',
+  // ID     : 'ChangeHistoryFacet',
+  '@UI.PartOfPreview': false
+}
+
+
+/**
+ * Returns an expression for the key of the given entity, which we can use as the right-hand-side of an ON condition.
+ */
+function entityKey4 (entity) {
+  const xpr = []
+  for (let k in entity.elements) {
+    const e = entity.elements[k]; if (!e.key) continue
+    if (xpr.length) xpr.push('||')
+    if (e.type === 'cds.Association') xpr.push({ ref: [k, e.keys?.[0]?.ref?.[0]] })
+    else xpr.push({ ref:[k] })
+  }
+  return xpr
 }
 
+
 // Unfold @changelog annotations in loaded model
-cds.on('loaded', m => {
+function enhanceModel (csn) {
+
+  if (!csn.definitions?.[changes.target]) return // no change tracking in this model
+  if (csn.meta?.[_enhanced]) return // already enhanced
 
-  // Get definitions from Dummy entity in our models
-  const { 'sap.changelog.aspect': aspect } = m.definitions; if (!aspect) return // some other model
-  const { '@UI.Facets': [facet], elements: { changes } } = aspect
-  changes.on.pop() // remove ID -> filled in below
-  
   // Process entities to define the relation
-  processEntities(m)
+  processEntities(csn) // REVISIT: why is that required ?!?
 
-  for (let name in m.definitions) {
-    const entity = m.definitions[name]
+  for (let name in csn.definitions) {
+
+    const entity = csn.definitions[name]
     if (isChangeTracked(entity)) {
 
-      // Determine entity keys
-      const keys = [], { elements: elms } = entity
-      for (let e in elms) {
-        if (elms[e].key) {
-          if (elms[e].type === 'cds.Association') {
-            keys.push({
-              e,
-              val: elms[e]
-            });
-          } else {
-            keys.push(e);
+      if (!entity['@changelog.disable_assoc']) { // REVISIT: why do we need that annotation?
```

Comment:
This is a flag to disable automatic association definition. This is for people who want to change-track things, but don't want the change-log to be attached to the entity itself. They rather build a central Admin UI to view all changes.
One of the reasons is also, that ALL changes are visible through this association, regardless of who did them, something some stakeholders see as security issue